features/253-auth-sdk-callback-in-next/api.md

```diff
@@ -25,233 +25,143 @@ skygear.publicDB.save(user).then((user) => {
+- auth info: any data that may affect or affected by user authentication status or authorized status, such as disabled, last login at, roles, ..., etc.
+- metadata: any user attributes that are not auth info specified, by default, it contains some pre-defined common user attributes, and a developer is allowed to add custom user attributes here .
+- user object: an object combines auth info and metadata and represents as an user.
 
-So in skygear next, even though record gear is optional, we won't introduce a new `coreUser` data type, on the contrary, we will keep return data type of auth gear as `record` to reduce any possible misunderstanding. 
+# Pre-defined common attributes
 
-For the case record gear is missing, auth gear will have a simple user profile implementatio by adding a json field to `_core_user` table. And it will also generate a record alike data arribute to client SDK.
-
-# Example usage on JS SDK
-
-keep current design, return data type is `record`.
-      
 ```
-+--------------------------+
-|                          |
-| [ currentUser <record> ] |
-| username                 |
-| email                    |
-| phone                    |
-| .....                    |
-|                          |
-+--------------------------+
-
-// user record
-skygear.auth.currentUser;
-
-// current user auth info
-skygear.auth.isCurrentUserVerified;
-skygear.auth.isCurrentUserDisabled;
-
-// current user profile record
-const phone = skygear.auth.currentUser['phone'];
-
-// update record
-skygear.auth.currentUser['username'] = 'new-username';
-skygear.auth.currentUser['phone'] = 'new-phone';
-skygear.auth.updateUser(skygear.auth.currentUser);
+avatar_url text
+name text
+nickname text
+birthday timestamp
+preferred_lang text
 ```
 
-# Changes on SDK
+- `preferred_lang` is a string in RFC 5646, ex: 'en', 'zh-TW', 'zh-CN', 'zh-HK'.
 
-Skygear next server will help generate `record` data type for SDK, therefore SDKs are expected no major changes.
+# User object response from backend
 
-Such changes are occurred in server side rather than SDK because:
-
-- SDK have zero knowledge about whether record configured or not.
-- Server side implementation would apply to multiple SDKs at once.
-- Server side implementation eliminates possible errors.
+```
+{
+    user_id: <id>,
+    created_at: <created_at>,
+    updated_at: <updated_at>,
+    last_seen_at: <last_seen_at>,
+    verified: <verified>,
+    verify_info: <verified_info>,
+    roles: [<role>, <role>, <role>, ...],
+    username: <username>,
+    email: <email>,
+    metadata: {
+        avatar_url: <avatarUrl>,
+        name: <name>,
+        nickname: <nickname>,
+        birthday: <birthday>,
+        preferred_lang: <preferred_lang>,
+        
+        // any other custom attributes
+        ...
```

Comment:
First thing want to confirm about `AUTH_RECORD_KEYS` and metadata. If developer configure other keys, e.g. phone, as `AUTH_RECORD_KEYS`, would that be a top level key? And if username and/or email are not in `AUTH_RECORD_KEYS`, are they part of predefined attributes in metadata?

Second, I think when considering what are predefined attributes, we need a real use case.
For example, avatar_url are commonly used when link with social account and maybe chat app. However, nickname and preferred_lang seems no real use case. 
It was mentioned before that these attributes are for user portal, however, if we now support custom attributes in meta data, and the support for both predefined and custom attributes should be the same, unless the list of predefined attributes is large enough that cover different area, which is pretty impossible. So I don't think user portal is a strong supporting use case when defining user portal.

Reply:
@Steven-Chan 

> First thing want to confirm about `AUTH_RECORD_KEYS` and metadata. If developer configure other keys, e.g. phone, as `AUTH_RECORD_KEYS`, would that be a top level key? And if username and/or email are not in `AUTH_RECORD_KEYS`, are they part of predefined attributes in metadata?

1. I consider `AUTH_RECORD_KEYS` will represent keys  at top leavel. For instance, if `AUTH_RECORD_KEYS=[["phone"]]`, then user object is:

   ```
   {
     user_id: <id>,
     created_at: <created_at>,
     updated_at: <updated_at>,
     last_seen_at: <last_seen_at>,
     verified: <verified>,
     verify_info: <verified_info>,
     roles: [<role>, <role>, <role>, ...],
     phone: "<a_phone_number>",
     metadata: {
         avatar_url: <avatarUrl>,
         name: <name>,
         nickname: <nickname>,
         birthday: <birthday>,
         preferred_lang: <preferred_lang>,
     }
   }
   ```

2. username/email will not at pre-defined user attributes, though app still can define username/email in custom user data in metadata, they can be treated as normal user attribute, like

   ```
   {
     ...
     phone: "<a_phone_number>",
     metadata: {
         avatar_url: <avatarUrl>,
         name: <name>,
         nickname: <nickname>,
         birthday: <birthday>,
         preferred_lang: <preferred_lang>,

         // custom user data
         username: <username>,
         email: <email>
     }
   }
   ```

> Second, I think when considering what are predefined attributes, we need a real use case.
> For example, avatar_url are commonly used when link with social account and maybe chat app. However, nickname and preferred_lang seems no real use case.
> It was mentioned before that these attributes are for user portal, however, if we now support custom attributes in meta data, and the support for both predefined and custom attributes should be the same, unless the list of predefined attributes is large enough that cover different area, which is pretty impossible. So I don't think user portal is a strong supporting use case when defining predefined attributes.

I can understand your concern; current "pre-defined user attributes" design base on a hypothesize: "We will have user-management dashboard in a short time" (https://github.com/SkygearIO/features/pull/261#issuecomment-454262346). And base on this, so we

1. manage user profile in auth gear
2. pre-defined user attributes

Though I do have troble to define a large enough attribute set to cover all possible use cases.